### PR TITLE
Status host uses net.port by default, add UPnP port forwarding option

### PR DESCRIPTION
**Pull request by 20kdc**
*CHECK/FIX PRODUCTION SERVER CONFIGS BEFORE THIS REACHES PRODUCTION!!!*
should be compatible but check anyway

UPnP is not enabled by default because it could mess with stuff that people don't want messed with

### CODE REVIEW

**Comment by PJB3005** [comment]
image Good tests, github

**Comment by 20kdc** [comment]
ok, so, if you look at the test failure details, the answer is that the criteria for a test failure is *any* warning
and *any* warning includes, say, a UPnP failure warning

File: Robust.Server/Network/UPnP.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            // We DON'T want to hold up the main server on this!
+            new Thread(() =>
+            {
+                try
+                {
+                    // The way the NetUPnP code is written, we're doing guesswork anyway
+                    // It seems to be that the assumption in regards to IPv6 is "what?" w/ UPnP????
+                    // ATTENTION FUTURE 20KDC (or anyone else who comes by):
+                    // IF YOU GET IPv6 FOR REALSIES, WORK OUT HOW TO DEAL W/ THIS!
```
Comment (PJB3005) [inline Robust.Server/Network/UPnP.cs]: IPv6 support was retrofitted into Lidgren (by me) so take that as you will.

Comment (20kdc) [reply Robust.Server/Network/UPnP.cs]: supposedly the IPv6 Plan (r) is "no NAT!" so we'll have to see how it goes

**Comment by 20kdc** [comment]
disabled UPnP by default for... well, various reasons really.

File: Robust.Server/Network/UPnP.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+using Robust.Shared.Log;
+
+namespace Robust.Server.Network
+{
+    internal static class UPnP
+    {
+        public static void Start(int port, ISawmill sawmill)
+        {
+            // We DON'T want to hold up the main server on this!
+            new Thread(() =>
```
Comment (PJB3005) [inline Robust.Server/Network/UPnP.cs]: Using `Task.Run()` here is probably easier?

Comment (20kdc) [reply Robust.Server/Network/UPnP.cs]: Does the thread-pool automatically expand if one of the threads blocks for a long period of time?

Comment (PJB3005) [reply Robust.Server/Network/UPnP.cs]: It does, yeah. But I figure the amount of time this takes shouldn't be too long to cause a problem.

The even higher effort approach would be to make the Lidgren APIs for this async but ah well. You can at least make the `NetUtility.Sleep` used to poll the UPnP status a `Task.Delay` instead (obviously ideally it'd be an event or smth but ah well...)

Comment (20kdc) [reply Robust.Server/Network/UPnP.cs]: ah, I was under the assumption that using NetUtility.Sleep was critically necessary to the functioning of Lidgren

Comment (PJB3005) [reply Robust.Server/Network/UPnP.cs]: It just calls `Thread.Sleep()` directly, so...

**Comment by mirrorcult** [comment]
should update the hosting docs once this is merged to indicate that you can use UPnP

**Comment by PJB3005** [comment]
Lidgren needs some work on its side to fix the UPnP stuff, working on it.

**Pull request merged**